ccl/multiregionccl: TestRegionLivenessProber failed

ccl/multiregionccl.TestRegionLivenessProber [failed](https://tanzanite.cluster.engflow.com/invocations/default/9ae3e053-71d9-44f9-90eb-f9ca6446aa4d?testReportRun=13&testReportShard=3&testReportAttempt=1#targets-Ly9wa2cvY2NsL211bHRpcmVnaW9uY2NsOm11bHRpcmVnaW9uY2NsX3Rlc3Q=) on release-25.1 @ [839543761f65547a48b78c10af993545590bded5](https://github.com/cockroachdb/cockroach/commits/839543761f65547a48b78c10af993545590bded5):


```
=== RUN   TestRegionLivenessProber
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestRegionLivenessProber1843916692
    test_log_scope.go:76: use -show-logs to present logs inline
    regionliveness_test.go:176: 
        	Error Trace:	pkg/ccl/multiregionccl/regionliveness_test.go:176
        	Error:      	Not equal: 
        	            	expected: 1
        	            	actual  : 0
        	Test:       	TestRegionLivenessProber
        	Messages:   	one region should be dead
    panic.go:626: -- test log scope end --
test logs left over in: outputs.zip/logTestRegionLivenessProber1843916692
--- FAIL: TestRegionLivenessProber (20.35s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=13</code>
 - <code>shard=3</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRegionLivenessProber.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-47302

--- a/pkg/ccl/multiregionccl/regionliveness_test.go
+++ b/pkg/ccl/multiregionccl/regionliveness_test.go
@@ -176,7 +176,9 @@ func TestRegionLivenessProber(t *testing.T) {
 			// Attempt to keep on pushing out the unavailable_at time,
 			// these calls should be no-ops.
 			blockProbeQuery.Store(true)
-			require.NoError(t, regionProber.ProbeLiveness(ctx, expectedRegions[1]))
+			if err := regionProber.ProbeLiveness(ctx, expectedRegions[1]); err != nil {
+				return err
+			}
 			// Check if the time has expired yet.
 			regions, err := regionProber.QueryLiveness(ctx, txn)
 			if err != nil {

--- a/pkg/ccl/multiregionccl/regionliveness_test.go
+++ b/pkg/ccl/multiregionccl/regionliveness_test.go
@@ -424,16 +424,16 @@ func TestRegionLivenessProberForLeases(t *testing.T) {
 	tr := sqlutils.MakeSQLRunner(newRegionSQL)
 	// Validate everything was cleaned bringing up a new node in the down region.
 	require.Equalf(t,
-		tr.QueryStr(t, "SELECT * FROM system.region_liveness"),
 		[][]string{},
+		tr.QueryStr(t, "SELECT * FROM system.region_liveness"),
 		"expected no unavaialble regions.")
 	require.Equalf(t,
-		tr.QueryStr(t, "SELECT count(*) FROM system.sql_instances WHERE session_id IS NOT NULL"),
 		[][]string{{"3"}},
+		tr.QueryStr(t, "SELECT count(*) FROM system.sql_instances WHERE session_id IS NOT NULL"),
 		"extra sql instances are being used.")
 	require.Equalf(t,
-		tr.QueryStr(t, "SELECT count(*) FROM system.sqlliveness"),
 		[][]string{{"3"}},
+		tr.QueryStr(t, "SELECT count(*) FROM system.sqlliveness"),
 		"extra sql sessions detected.")
 	require.NoError(t, err)
 	// Validate that the stuck query will fail once we recover.

--- a/pkg/ccl/multiregionccl/regionliveness_test.go
+++ b/pkg/ccl/multiregionccl/regionliveness_test.go
@@ -100,7 +100,7 @@ func TestRegionLivenessProber(t *testing.T) {
 	var tenants []serverutils.ApplicationLayerInterface
 	var tenantSQL []*gosql.DB
 	blockProbeQuery := atomic.Bool{}
-	defer regionliveness.TestingSetProbeLivenessTimeout(
+	defer regionliveness.TestingSetBeforeProbeLivenessHook(
 		func() {
 			// Timeout attempts to probe intentionally.
 			if blockProbeQuery.Swap(false) {
@@ -254,7 +254,7 @@ func TestRegionLivenessProberForLeases(t *testing.T) {
 	targetCount := atomic.Int64{}
 	var tenants []serverutils.ApplicationLayerInterface
 	var tenantSQL []*gosql.DB
-	defer regionliveness.TestingSetProbeLivenessTimeout(
+	defer regionliveness.TestingSetBeforeProbeLivenessHook(
 		func() {
 			if !detectLeaseWait.Load() {
 				return
@@ -522,7 +522,7 @@ func TestRegionLivenessProberForSQLInstances(t *testing.T) {
 	var tenants []serverutils.ApplicationLayerInterface
 	var tenantSQL []*gosql.DB
 	blockProbeQuery := atomic.Bool{}
-	defer regionliveness.TestingSetProbeLivenessTimeout(
+	defer regionliveness.TestingSetBeforeProbeLivenessHook(
 		func() {
 			// Timeout attempts to probe intentionally.
 			if blockProbeQuery.Swap(false) {

--- a/pkg/ccl/multiregionccl/regionliveness_test.go
+++ b/pkg/ccl/multiregionccl/regionliveness_test.go
@@ -51,7 +51,7 @@ import (
 const (
 	// Use a low probe timeout number, but intentionally only manipulate to
 	// this value when *failures* are expected.
-	testingRegionLivenessProbeTimeout = time.Second * 2
+	testingRegionLivenessProbeTimeout = time.Second * 1
 	// Use a long probe timeout by default when running this test (since CI
 	// can easily hit query timeouts).
 	testingRegionLivenessProbeTimeoutLong = time.Minute
@@ -104,9 +104,10 @@ func TestRegionLivenessProber(t *testing.T) {
 		func() {
 			// Timeout attempts to probe intentionally.
 			if blockProbeQuery.Swap(false) {
-				time.Sleep(testingRegionLivenessProbeTimeout)
+				time.Sleep(2 * testingRegionLivenessProbeTimeout)
 			}
-		})()
+		},
+	)()
 
 	for _, s := range testCluster.Servers {
 		tenantArgs := base.TestTenantArgs{
@@ -253,14 +254,16 @@ func TestRegionLivenessProberForLeases(t *testing.T) {
 	targetCount := atomic.Int64{}
 	var tenants []serverutils.ApplicationLayerInterface
 	var tenantSQL []*gosql.DB
-	defer regionliveness.TestingSetProbeLivenessTimeout(func() {
-		if !detectLeaseWait.Load() {
-			return
-		}
-		time.Sleep(testingRegionLivenessProbeTimeout)
-		targetCount.Swap(0)
-		detectLeaseWait.Swap(false)
-	})()
+	defer regionliveness.TestingSetProbeLivenessTimeout(
+		func() {
+			if !detectLeaseWait.Load() {
+				return
+			}
+			time.Sleep(2 * testingRegionLivenessProbeTimeout)
+			targetCount.Swap(0)
+			detectLeaseWait.Swap(false)
+		},
+	)()
 
 	var keyToBlockMu syncutil.Mutex
 	var keyToBlock roachpb.Key
@@ -321,7 +324,7 @@ func TestRegionLivenessProberForLeases(t *testing.T) {
 							keyToBlockMu.Lock()
 							keyToBlock = tenant.Codec().TablePrefix(uint32(systemschema.RegionLivenessTable.GetID()))
 							keyToBlockMu.Unlock()
-							time.Sleep(testingRegionLivenessProbeTimeout)
+							time.Sleep(2 * testingRegionLivenessProbeTimeout)
 						}
 					},
 				},
@@ -523,9 +526,10 @@ func TestRegionLivenessProberForSQLInstances(t *testing.T) {
 		func() {
 			// Timeout attempts to probe intentionally.
 			if blockProbeQuery.Swap(false) {
-				time.Sleep(testingRegionLivenessProbeTimeout)
+				time.Sleep(2 * testingRegionLivenessProbeTimeout)
 			}
-		})()
+		},
+	)()
 
 	for _, s := range testCluster.Servers {
 		tenantArgs := base.TestTenantArgs{

--- a/pkg/ccl/multiregionccl/regionliveness_test.go
+++ b/pkg/ccl/multiregionccl/regionliveness_test.go
@@ -463,7 +463,6 @@ func TestRegionLivenessProberForLeases(t *testing.T) {
 			expectedTxnErr,
 			"txn should see a retry error")
 	}
-	require.ErrorContainsf(t, grp.Wait(), "context canceled", "connection should have been dropped, node is dead.")
 }
 
 // TestRegionLivenessProberForSQLInstances validates that regional avaibility issues

--- a/pkg/ccl/multiregionccl/regionliveness_test.go
+++ b/pkg/ccl/multiregionccl/regionliveness_test.go
@@ -406,41 +406,48 @@ func TestRegionLivenessProberForLeases(t *testing.T) {
 		regionliveness.RegionLivenessProbeTimeout.Override(ctx, &ts.ClusterSettings().SV, testingRegionLivenessProbeTimeoutLong)
 	}
 	require.NoError(t, lm.WaitForNoVersion(ctx, descpb.ID(tableID), cachedDatabaseRegions, retry.Options{}))
+
+	// Use a closure so that we unconditionally wait for the recovery to finish.
 	grp := ctxgroup.WithContext(ctx)
-	grp.GoCtx(func(ctx context.Context) error {
-		_, err = tx.Exec("INSERT INTO t2 VALUES(5)")
-		return err
-	})
-	// Add a new region which will execute a recovery and clean up dead rows.
-	keyToBlockMu.Lock()
-	keyToBlock = nil
-	keyToBlockMu.Unlock()
-	tenantArgs := base.TestTenantArgs{
-		Settings: makeSettings(),
-		TenantID: id,
-		Locality: testCluster.Servers[0].Locality(),
-	}
-	_, newRegionSQL := serverutils.StartTenant(t, testCluster.Servers[0], tenantArgs)
-	tr := sqlutils.MakeSQLRunner(newRegionSQL)
-	// Validate everything was cleaned bringing up a new node in the down region.
-	require.Equalf(t,
-		[][]string{},
-		tr.QueryStr(t, "SELECT * FROM system.region_liveness"),
-		"expected no unavaialble regions.")
-	require.Equalf(t,
-		[][]string{{"3"}},
-		tr.QueryStr(t, "SELECT count(*) FROM system.sql_instances WHERE session_id IS NOT NULL"),
-		"extra sql instances are being used.")
-	require.Equalf(t,
-		[][]string{{"3"}},
-		tr.QueryStr(t, "SELECT count(*) FROM system.sqlliveness"),
-		"extra sql sessions detected.")
-	require.NoError(t, err)
+	func() {
+		grp.GoCtx(func(ctx context.Context) error {
+			_, err = tx.Exec("INSERT INTO t2 VALUES(5)")
+			return err
+		})
+		defer func() {
+			<-recoveryStart
+			time.Sleep(slbase.DefaultTTL.Default())
+			recoveryBlock <- struct{}{}
+			require.NoError(t, grp.Wait())
+		}()
+		// Add a new region which will execute a recovery and clean up dead rows.
+		keyToBlockMu.Lock()
+		keyToBlock = nil
+		keyToBlockMu.Unlock()
+		tenantArgs := base.TestTenantArgs{
+			Settings: makeSettings(),
+			TenantID: id,
+			Locality: testCluster.Servers[0].Locality(),
+		}
+		_, newRegionSQL := serverutils.StartTenant(t, testCluster.Servers[0], tenantArgs)
+		tr := sqlutils.MakeSQLRunner(newRegionSQL)
+		// Validate everything was cleaned bringing up a new node in the down region.
+		require.Equalf(t,
+			[][]string{},
+			tr.QueryStr(t, "SELECT * FROM system.region_liveness"),
+			"expected no unavaialble regions.")
+		require.Equalf(t,
+			[][]string{{"3"}},
+			tr.QueryStr(t, "SELECT count(*) FROM system.sql_instances WHERE session_id IS NOT NULL"),
+			"extra sql instances are being used.")
+		require.Equalf(t,
+			[][]string{{"3"}},
+			tr.QueryStr(t, "SELECT count(*) FROM system.sqlliveness"),
+			"extra sql sessions detected.")
+		require.NoError(t, err)
+	}()
+
 	// Validate that the stuck query will fail once we recover.
-	<-recoveryStart
-	time.Sleep(slbase.DefaultTTL.Default())
-	recoveryBlock <- struct{}{}
-	require.NoError(t, grp.Wait())
 	_, err = tx.Exec("INSERT INTO t2 VALUES(5)")
 	// If the txn failed, no commit is needed.
 	const expectedTxnErr = "restart transaction: TransactionRetryWithProtoRefreshError"

--- a/pkg/sql/regionliveness/prober.go
+++ b/pkg/sql/regionliveness/prober.go
@@ -113,13 +113,13 @@ type livenessProber struct {
 	settings        *clustersettings.Settings
 }
 
-var testingProbeQueryCallbackFunc func()
+var testingBeforeProbeLivenessHook func()
 var testingUnavailableAtTTLOverride time.Duration
 
-func TestingSetProbeLivenessTimeout(probeCallbackFn func()) func() {
-	testingProbeQueryCallbackFunc = probeCallbackFn
+func TestingSetBeforeProbeLivenessHook(hook func()) func() {
+	testingBeforeProbeLivenessHook = hook
 	return func() {
-		testingProbeQueryCallbackFunc = nil
+		testingBeforeProbeLivenessHook = nil
 	}
 }
 
@@ -183,8 +183,8 @@ func (l *livenessProber) ProbeLivenessWithPhysicalRegion(
 	err := timeutil.RunWithTimeout(ctx, "probe-liveness", tableTimeout,
 		func(ctx context.Context) error {
 			return l.db.Txn(ctx, func(ctx context.Context, txn *kv.Txn) error {
-				if testingProbeQueryCallbackFunc != nil {
-					testingProbeQueryCallbackFunc()
+				if testingBeforeProbeLivenessHook != nil {
+					testingBeforeProbeLivenessHook()
 				}
 				instancesTable := systemschema.SQLInstancesTable()
 				indexPrefix := l.codec.IndexPrefix(uint32(instancesTable.GetID()), uint32(instancesTable.GetPrimaryIndexID()))

--- a/pkg/sql/regionliveness/prober.go
+++ b/pkg/sql/regionliveness/prober.go
@@ -119,7 +119,7 @@ var testingUnavailableAtTTLOverride time.Duration
 func TestingSetProbeLivenessTimeout(probeCallbackFn func()) func() {
 	testingProbeQueryCallbackFunc = probeCallbackFn
 	return func() {
-		probeCallbackFn = nil
+		testingProbeQueryCallbackFunc = nil
 	}
 }
 

